File: code_samples/section6/problems/problem6_3/problem6_3.cs
/*
    This file demonstrates how to find the k smallest elements in an array
    using a MAX-HEAP implemented via .NET's PriorityQueue<TElement, TPriority>.

    Core idea:
      - Maintain a max-heap of size at most k.
      - The heap always contains the k smallest elements seen so far.
      - The largest of those k elements sits at the root (Peek()).
      - When a new element x arrives:
          * If heap has fewer than k elements → insert x.
          * Else if x < current maximum → remove max, insert x.
          * Else → ignore x.

    Time complexity:
      - O(n log k), where n = arr.Length
    Space complexity:
      - O(k)

    Important edge cases:
      - k <= 0 → return empty list
      - k > n  → return all elements
      - arr is null or empty → return empty list
*/

/*
    KSmallest
    ---------
    Returns the k smallest elements from arr using a max-heap.

    NOTE:
      - This method assumes k > 0 and arr is non-null.
      - KSmallestSafe() wraps this method to handle edge cases.
*/
static IList<int> KSmallest(int[] arr, int k) {

    /*
        PriorityQueue<TElement, TPriority> is a MIN-HEAP by default.
        To turn it into a MAX-HEAP, we reverse the comparison:
            b.CompareTo(a)
        This causes larger integers to be treated as "smaller" priorities,
        making Peek() return the maximum element.
    */
    var pq = new PriorityQueue<int, int>(
        Comparer<int>.Create((a, b) => b.CompareTo(a))  // max-heap behavior
    );

    /*
        Iterate through the input array and maintain a heap of size <= k.
    */
    foreach (var x in arr) {

        if (pq.Count < k) {
            // Heap not full yet: just insert the element.
            // Element and priority are both x.
            pq.Enqueue(x, x);

        } else if (x < pq.Peek()) {
            /*
                Heap already has k elements.
                pq.Peek() returns the LARGEST of the k smallest elements so far.
                If x is smaller than that, x belongs in the k smallest set.
            */
            pq.Dequeue();       // Remove the current largest
            pq.Enqueue(x, x);   // Insert the smaller value
        }

        // Else:
        //   x is >= pq.Peek(), so it cannot be among the k smallest elements.
        //   Ignore it.
    }

    /*
        Extract elements from the heap.
        Since this is a max-heap, elements will come out in descending order.
        Order is not guaranteed or required by the problem.
    */
    var result = new List<int>();
    while (pq.Count > 0) {
        result.Add(pq.Dequeue());
    }

    return result;
}

/*
    KSmallestSafe
    -------------
    A defensive wrapper around KSmallest that handles:
      - k <= 0
      - null array
      - empty array

    Returns an empty list in those cases.
*/
static IList<int> KSmallestSafe(int[] arr, int k) {
    if (k <= 0 || arr == null || arr.Length == 0)
        return [];

    return KSmallest(arr, k);
}

// --------- Helpers ---------

/*
    PrintArray
    ----------
    Prints an integer array with a label.
*/
static void PrintArray(string label, int[] arr) {
    Console.Write(label);
    foreach (var x in arr) Console.Write(x + " ");
    Console.WriteLine();
}

/*
    PrintList
    ---------
    Prints any IEnumerable<int> with a label.
*/
static void PrintList(string label, IEnumerable<int> list) {
    Console.Write(label);
    foreach (var x in list) Console.Write(x + " ");
    Console.WriteLine();
}

/*
    TestKSmallest
    -------------
    Test harness that exercises KSmallestSafe with:
      1) k <= n
      2) k > n
      3) k = 0

    For readability, results are also sorted after extraction
    so correctness is easy to visually verify.
*/
static void TestKSmallest() {
    var arr = new[] { 7, 2, 9, 4, 1, 5, 8, 3, 6 };

    // ---- Test 1: k <= n ----
    Console.WriteLine("=== Test 1: k <= n ===");
    PrintArray("Input array: ", arr);

    var k = 4;
    Console.WriteLine("k = " + k);

    var res1 = KSmallestSafe(arr, k);
    PrintList("KSmallest result (heap order): ", res1);

    // Sort for easier verification
    var sorted1 = new List<int>(res1);
    sorted1.Sort();
    PrintList("KSmallest result (sorted): ", sorted1);
    Console.WriteLine("(Expected: 1 2 3 4)");
    Console.WriteLine();

    // ---- Test 2: k > n ----
    Console.WriteLine("=== Test 2: k > n ===");
    k = 20;
    PrintArray("Input array: ", arr);
    Console.WriteLine("k = " + k);

    var res2 = KSmallestSafe(arr, k);
    PrintList("KSmallest result (heap order): ", res2);

    var sorted2 = new List<int>(res2);
    sorted2.Sort();
    PrintList("KSmallest result (sorted): ", sorted2);
    Console.WriteLine("(Expected: all elements sorted)");
    Console.WriteLine();

    // ---- Test 3: k = 0 ----
    Console.WriteLine("=== Test 3: k = 0 ===");
    k = 0;
    Console.WriteLine("k = " + k);

    var res3 = KSmallestSafe(arr, k);
    PrintList("KSmallest result: ", res3);
    Console.WriteLine("(Expected: empty list)");
    Console.WriteLine();
}

// Execute the test harness
TestKSmallest();

File: code_samples/section6/problems/problem6_3/problem6_3.java
package code_samples.section6.problems.problem6_3;

import java.util.*;

/*
    Problem 6.3 - Find the k smallest elements using a max-heap (Java).

    Core idea:
      - Maintain a MAX-HEAP of size at most k.
      - The heap always contains the k smallest elements seen so far.
      - The root of the max-heap is the *largest* among those k smallest.
      - When a new element x arrives:
          * If the heap has fewer than k elements → insert x.
          * Else if x < heap.peek() → remove the largest and insert x.
          * Else → ignore x.

    Why a max-heap?
      - We want fast access to the largest of the kept k elements.
      - That element defines the cutoff for whether a new value belongs
        among the k smallest.

    Complexity:
      - Time:  O(n log k), where n = arr.length
      - Space: O(k)

    Edge cases handled:
      - k <= 0
      - null or empty input array
*/

public class problem6_3 {

    /*
        kSmallest
        ---------
        Returns a list containing the k smallest elements of the input array.

        Parameters:
            arr : input array of integers
            k   : number of smallest elements to return

        Returns:
            - A list containing up to k smallest elements (unordered).
            - Returns an empty list if k <= 0, arr is null, or arr is empty.

        Notes:
            - The returned list is NOT sorted.
            - Because a max-heap is used, iteration order typically reflects
              heap order (not ascending).
    */
    List<Integer> kSmallest(int[] arr, int k) {

        // Handle edge cases up front to keep main logic simple
        if (k <= 0 || arr == null || arr.length == 0) {
            return new ArrayList<>();
        }

        /*
            Java's PriorityQueue is a MIN-HEAP by default.
            To turn it into a MAX-HEAP, we reverse the comparator.

            Comparator explanation:
                (a, b) -> b - a

            This causes larger values to be considered "smaller" in priority,
            making peek() return the maximum element.
        */
        PriorityQueue<Integer> pq =
            new PriorityQueue<>((a, b) -> b - a); // max-heap

        /*
            Iterate through all elements in the array and maintain a heap
            of at most k elements.
        */
        for (int x : arr) {

            if (pq.size() < k) {
                // Heap not full yet: just add the element.
                pq.add(x);

            } else if (x < pq.peek()) {
                /*
                    Heap already contains k elements.

                    pq.peek() is the LARGEST of the k smallest elements so far.
                    If x is smaller than that, x belongs in the result set.
                */
                pq.poll();   // remove current largest
                pq.add(x);   // insert the smaller element
            }

            // Else:
            //   x >= pq.peek(), so x cannot be among the k smallest elements.
            //   Ignore it.
        }

        /*
            Convert heap contents to a list.

            Important:
              - Order is arbitrary (heap order), NOT sorted.
              - Sorting can be done by the caller if required.
        */
        return new ArrayList<>(pq);
    }

    // ===== Test Harness =====

    /*
        printArray
        ----------
        Utility function to print an int[] with a label.
    */
    private static void printArray(String label, int[] arr) {
        System.out.print(label);
        for (int x : arr) System.out.print(x + " ");
        System.out.println();
    }

    /*
        printList
        ---------
        Utility function to print a List<Integer> with a label.
    */
    private static void printList(String label, List<Integer> list) {
        System.out.print(label);
        for (int x : list) System.out.print(x + " ");
        System.out.println();
    }

    /*
        main
        ----
        Test driver that exercises kSmallest with:
          1) k <= n
          2) k > n
          3) k = 0

        For verification:
          - Results are also sorted before printing expected output,
            since kSmallest does not guarantee order.
    */
    public static void main(String[] args) {
        problem6_3 solver = new problem6_3();

        int[] arr = {7, 2, 9, 4, 1, 5, 8, 3, 6};

        // ---- Test 1: k <= n ----
        System.out.println("=== Test 1: k <= n ===");
        printArray("Input array: ", arr);

        int k = 4;
        System.out.println("k = " + k);

        List<Integer> result1 = solver.kSmallest(arr, k);
        printList("kSmallest result (heap order): ", result1);

        // Sort result for easy correctness checking
        List<Integer> sorted1 = new ArrayList<>(result1);
        Collections.sort(sorted1);
        printList("kSmallest result (sorted): ", sorted1);
        System.out.println("(Expected: 1 2 3 4)\n");

        // ---- Test 2: k > n ----
        System.out.println("=== Test 2: k > n ===");
        k = 20;

        printArray("Input array: ", arr);
        System.out.println("k = " + k);

        List<Integer> result2 = solver.kSmallest(arr, k);
        printList("kSmallest result (heap order): ", result2);

        List<Integer> sorted2 = new ArrayList<>(result2);
        Collections.sort(sorted2);
        printList("kSmallest result (sorted): ", sorted2);
        System.out.println("(Expected: all elements sorted)\n");

        // ---- Test 3: k = 0 ----
        System.out.println("=== Test 3: k = 0 ===");
        k = 0;

        System.out.println("k = " + k);
        List<Integer> result3 = solver.kSmallest(arr, k);
        printList("kSmallest result: ", result3);
        System.out.println("(Expected: empty list)\n");
    }
}
